### PR TITLE
Fixes to onboarding `Cluster`, `IPAddress`, `Prefix`, `Rack` to Dynamic Groups

### PR DESCRIPTION
**Pull request by jathanism**
# Replaces: #2188 
# What's Changed
- `DynamicGroup._map_filter_fields` now explicitly skips any method filters. This is a stop-gap until we are able to vet every method filter on every filterset. This is because we moved to Q object filtering for nested groups.
- Introduced a new pattern for filterset method filters to be backed by a corresponding `generate_query_{filter_method}` method that itself emits a `Q` object that can be used by `DynamicGroup.generate_query()` machinery
 - Specific filterset fields have been revised to have these `generate_query_` filterset methods defined to maintain feature parity w/ v1.3 Dynamic Groups (in the case of Device/VM), and list view filtering (in the case of Rack, IPAddress, Prefix)
- Fixed a bug in which method filters would only be filtered out if they were found to be a missing field
# TODO

- [x] Explanation of Change(s)
- [ ] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/ipam/filters.py [hunk 1]
```diff
@@ -316,10 +316,19 @@ def search_contains(self, queryset, name, value):
         except (AddrFormatError, ValueError):
             return queryset.none()
 
+    def generate_query_filter_present_in_vrf(self, value):
+        if isinstance(value, str):
+            value = VRF.objects.get(pk=value)
+
+        # This stringification is done to make the `pretty_print_query()` output look human-readable,
+        # and nothing more. It would work as complex objects but looks bad in the web UI.
```
Comment (glennmatthews) [inline nautobot/ipam/filters.py:324]: Is this still an issue after #2197?

File: nautobot/extras/tests/test_dynamicgroups.py [hunk 2]
```diff
@@ -346,6 +347,50 @@ def test_map_filter_fields_skip_missing(self):
...
+        # Make some field presence assertions
+        self.assertIn(field_name, fields)
+
+        # The filterset should have the method name and `generate_query_` method
+        self.assertTrue(hasattr(filterset, filter_field.method))
+        self.assertTrue(hasattr(filterset, "generate_query_" + filter_field.method))
+
+    # 2.0 TODO(jathan): This is done using Prefix at this time and this should be revised as filter
+    # fields are vetted.
+    def tets_map_filter_fields_skip_method_filters_no_generate_query(self):
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_dynamicgroups.py:372]:
```suggestion
    def test_map_filter_fields_skip_method_filters_no_generate_query(self):
```

File: nautobot/extras/tests/test_dynamicgroups.py [hunk 3]
```diff
@@ -346,6 +347,50 @@ def test_map_filter_fields_skip_missing(self):
...
+        # We know that `has_primary_ip` fits this bill, so let's test that.
+        field_name = "has_primary_ip"
+        filter_field = filterset.filters[field_name]
+
+        # Make some field presence assertions
+        self.assertIn(field_name, fields)
+
+        # The filterset should have the method name and `generate_query_` method
+        self.assertTrue(hasattr(filterset, filter_field.method))
+        self.assertTrue(hasattr(filterset, "generate_query_" + filter_field.method))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_dynamicgroups.py:368]: Where are we testing the query construction that involves the `generate_query_*` methods?

Comment (jathanism) [reply nautobot/extras/tests/test_dynamicgroups.py:368]: We aren't since it's going to vary wildly by method. Are you thinking we should test at least one that we know works a certain way to assert it?

Comment (glennmatthews) [reply nautobot/extras/tests/test_dynamicgroups.py:368]: Yeah, testing that the dynamic group correctly constructs the appropriate query for a given set of filters that rely on this feature.

File: nautobot/ipam/filters.py (hunk 1 above)
Comment (jathanism) [reply nautobot/ipam/filters.py:324]: Yes:

image

Compared to your fixes with leaving the stringification in there:

image

Comment (glennmatthews) [reply nautobot/ipam/filters.py:324]: Cool, thanks. Not blocking but wonder if we can/should fix this in `pretty_print_query()` itself?

Comment (jathanism) [reply nautobot/ipam/filters.py:324]: Yeah I think that is probably better so it's not up to plugin authors or us in the future.

**Comment by mzbroch** [comment]
Thanks Jathan! User experience looks much better now!

File: nautobot/dcim/models/racks.py [hunk 4]
```diff
@@ -310,6 +311,10 @@ class Rack(PrimaryModel, StatusModel):
         "outer_depth",
         "outer_unit",
     ]
+    dynamic_group_filter_fields = {
```
Comment (jathanism) [inline nautobot/dcim/models/racks.py:314]: Comment: Why is this necessary? Explain similarly to how it is done on `VirtualMachine`. And why are we still skipping fields?

File: nautobot/ipam/models.py [hunk 5]
```diff
@@ -524,6 +525,9 @@ class Prefix(PrimaryModel, StatusModel):
         "is_pool",
         "description",
     ]
+    dynamic_group_filter_fields = {
```
Comment (jathanism) [inline nautobot/ipam/models.py:528]: Comment: Why is this necessary? Explain similarly to how it is done on `VirtualMachine`.

File: nautobot/ipam/models.py [hunk 6]
```diff
@@ -839,6 +844,7 @@ class IPAddress(PrimaryModel, StatusModel):
         "role",
         "description",
     ]
+    dynamic_group_skip_missing_fields = True
```
Comment (jathanism) [inline nautobot/ipam/models.py:847]: Comment: Why are we skipping missing fields here?

File: nautobot/dcim/models/racks.py [hunk 7]
```diff
@@ -310,6 +311,10 @@ class Rack(PrimaryModel, StatusModel):
...
+        "group": "group_id",
+    }
+    dynamic_group_skip_missing_fields = True
```
Comment (bryanculver) [inline nautobot/dcim/models/racks.py:317]:
```suggestion
    dynamic_group_filter_fields = {
        "group": "group_id", # Duplicate filter fields that will be collapsed in 2.0
    }
    dynamic_group_skip_missing_fields = True # Poor widget selection for `outer_depth` (no validators, limit supplied)
```

File: nautobot/ipam/models.py [hunk 8]
```diff
@@ -524,6 +525,9 @@ class Prefix(PrimaryModel, StatusModel):
         "is_pool",
         "description",
     ]
+    dynamic_group_filter_fields = {
+        "vrf": "vrf_id",
+    }
```
Comment (bryanculver) [inline nautobot/ipam/models.py:530]:
```suggestion
    dynamic_group_filter_fields = {
        "vrf": "vrf_id", # Duplicate filter fields that will be collapsed in 2.0
    }
```

File: nautobot/ipam/models.py (hunk 6 above)
Comment (bryanculver) [inline nautobot/ipam/models.py:847]:
```suggestion
    dynamic_group_skip_missing_fields = True # Problematic form labels for `vminterface` and `interface`
```

**Comment by bryanculver** [comment]
Note here: We skipped automatic field population on a few of these models due in part to this: https://github.com/nautobot/nautobot/issues/2162 (ex: `rack_depth` on Rack)
